lib/srv/server/ec2_watcher.go

```diff
@@ -510,26 +608,36 @@ func (f *ec2InstanceFetcher) GetInstances(ctx context.Context, rotation bool) ([
 	f.cachedInstances.clear()
 	var allInstances []Instances
 
-	awsOpts := []awsconfig.OptionsFn{
-		awsconfig.WithCredentialsMaybeIntegration(awsconfig.IntegrationMetadata{Name: f.Matcher.Integration}),
-	}
-
-	if f.Matcher.AssumeRole != nil {
-		awsOpts = append(awsOpts, awsconfig.WithAssumeRole(f.Matcher.AssumeRole.RoleARN, f.Matcher.AssumeRole.ExternalID))
-	}
-
-	regions, err := f.matcherRegions(ctx, awsOpts)
+	accountRolesToAssume, err := f.allAssumeRoles(ctx)
 	if err != nil {
 		return nil, trace.Wrap(err)
 	}
 
-	for _, region := range regions {
-		regionInstances, err := f.getInstancesInRegion(ctx, rotation, region, awsOpts, ssmRunParams)
+	for _, assumeRole := range accountRolesToAssume {
+		awsOpts := []awsconfig.OptionsFn{
+			awsconfig.WithCredentialsMaybeIntegration(awsconfig.IntegrationMetadata{Name: f.Matcher.Integration}),
```

Comment:
Any single error in role assumption will break the entire process. We should allow for partial failures here or it will be very fragile in larger orgs.